Sources/NIO/DatagramVectorReadManager.swift

```diff
@@ -0,0 +1,203 @@
+            return try socket.recvmmsg(msgs: self.messageVector)
+        }
+
+        switch result {
+        case .wouldBlock(let messagesProcessed):
+            assert(messagesProcessed == 0)
+            return .none
+        case .processed(let messagesProcessed):
+            return self.buildMessages(messageCount: messagesProcessed, sliceSize: messageSize, buffer: &buffer)
+        }
+    }
+
+    /// Destroys this vector read manager, rendering it impossible to re-use. Use of the vector read manager after this is called is not possible.
+    mutating func destroy() {
+        self.messageVector.deinitializeAndDeallocate()
+        self.ioVector.deinitializeAndDeallocate()
+        self.sockaddrVector.deinitializeAndDeallocate()
+    }
+
+    private func buildMessages(messageCount: Int, sliceSize: Int, buffer: inout ByteBuffer) -> ReadResult {
+        var sliceOffset = buffer.readerIndex
+        var totalReadSize = 0
+
+        var results = Array<AddressedEnvelope<ByteBuffer>>()
+        results.reserveCapacity(messageCount)
+
+        for i in 0..<messageCount {
+            // We force-unwrap here because we should not have been able to write past the end of the buffer.
+            var slice = buffer.getSlice(at: sliceOffset, length: sliceSize)!
+            sliceOffset += sliceSize
+
+            // Now we move the writer index backwards to where the end of the read was. Note that 0 is not an
+            // error for datagrams, zero-length datagrams are permitted.
+            let readBytes = Int(self.messageVector[i].msg_len)
+            slice.moveWriterIndex(to: readBytes)
+            totalReadSize += readBytes
+
+            // Next we extract the remote peer address.
+            precondition(self.messageVector[i].msg_hdr.msg_namelen != 0, "Unexpected zero length peer name")
+            let address: SocketAddress = self.sockaddrVector[i].convert()
+
+            // Now we've finally constructed a useful AddressedEnvelope. We can store it in the results array temporarily.
+            results.append(AddressedEnvelope(remoteAddress: address, data: slice))
+        }
+
+        // Ok, all built. Now we can return these values to the caller.
+        return .some(reads: results, totalSize: totalReadSize)
+    }
+}
+
+
+extension Optional where Wrapped == DatagramVectorReadManager {
+    /// Updates the message count of the wrapped `DatagramVectorReadManager` to the new value.
+    ///
+    /// If the new value is 0 or negative, will destroy the contained vector read manager and set
+    /// self to `nil`.
+    mutating func updateMessageCount(_ newCount: Int) {
+        if newCount > 0 {
+            if self != nil {
+                self!.messageCount = newCount
+            } else {
+                self = DatagramVectorReadManager(messageCount: newCount)
+            }
+        } else {
+            if self != nil {
+                self!.destroy()
```

Comment:
here it's non-obvious to me why every `.init` is paired with exactly one `destroy`